registry/repository.go

```diff
@@ -69,42 +81,59 @@ func (c *repositoryClient) PushSignature(ctx context.Context, mediaType string,
 	}
 
 	return blobDesc, manifestDesc, nil
 }
 
-// getSignatureManifest returns signature manifest given signature manifest
-// descriptor
-func (c *repositoryClient) getSignatureManifest(ctx context.Context, sigManifestDesc ocispec.Descriptor) (*ocispec.Artifact, error) {
-	if sigManifestDesc.MediaType != ocispec.MediaTypeArtifactManifest {
-		return nil, fmt.Errorf("sigManifestDesc.MediaType requires %q, got %q", ocispec.MediaTypeArtifactManifest, sigManifestDesc.MediaType)
+// getSignatureBlobDesc returns signature blob descriptor from
+// signature manifest blobs or layers given signature manifest descriptor
+func (c *repositoryClient) getSignatureBlobDesc(ctx context.Context, sigManifestDesc ocispec.Descriptor) (ocispec.Descriptor, error) {
+	if sigManifestDesc.MediaType != ocispec.MediaTypeArtifactManifest && sigManifestDesc.MediaType != ocispec.MediaTypeImageManifest {
+		return ocispec.Descriptor{}, fmt.Errorf("sigManifestDesc.MediaType requires %q or %q, got %q", ocispec.MediaTypeArtifactManifest, ocispec.MediaTypeImageManifest, sigManifestDesc.MediaType)
 	}
 	if sigManifestDesc.Size > maxManifestSizeLimit {
-		return nil, fmt.Errorf("manifest too large: %d bytes", sigManifestDesc.Size)
+		return ocispec.Descriptor{}, fmt.Errorf("signature manifest too large: %d bytes", sigManifestDesc.Size)
 	}
 	manifestJSON, err := content.FetchAll(ctx, c.Repository.Manifests(), sigManifestDesc)
 	if err != nil {
-		return nil, err
+		return ocispec.Descriptor{}, err
 	}
 
-	var sigManifest ocispec.Artifact
-	err = json.Unmarshal(manifestJSON, &sigManifest)
-	if err != nil {
-		return nil, err
+	var signatureBlobs []ocispec.Descriptor
+
+	// OCI image manifest
+	if sigManifestDesc.MediaType == ocispec.MediaTypeImageManifest {
+		var sigManifest ocispec.Manifest
+		if err := json.Unmarshal(manifestJSON, &sigManifest); err != nil {
+			return ocispec.Descriptor{}, err
+		}
+		signatureBlobs = sigManifest.Layers
+	} else { // OCI artifact manifest
+		var sigManifest ocispec.Artifact
+		if err := json.Unmarshal(manifestJSON, &sigManifest); err != nil {
+			return ocispec.Descriptor{}, err
+		}
+		signatureBlobs = sigManifest.Blobs
+	}
+
+	if len(signatureBlobs) != 1 {
+		return ocispec.Descriptor{}, fmt.Errorf("signature manifest requries exactly one signature envelope blob, got %d", len(signatureBlobs))
 	}
-	return &sigManifest, nil
+
+	return signatureBlobs[0], nil
 }
 
 // uploadSignatureManifest uploads the signature manifest to the registry
-func (c *repositoryClient) uploadSignatureManifest(ctx context.Context, subject, blobDesc ocispec.Descriptor, annotations map[string]string) (ocispec.Descriptor, error) {
+func (c *repositoryClient) uploadSignatureManifest(ctx context.Context, subject, blobDesc ocispec.Descriptor, annotations map[string]string, ociImageManifest bool) (ocispec.Descriptor, error) {
 	opts := oras.PackOptions{
 		Subject:             &subject,
 		ManifestAnnotations: annotations,
+		PackImageManifest:   ociImageManifest,
```

Comment:
`ociImageManifest` can be accessed directly from `c.OCIImageManifest` instead of the function parameter.

Reply:
Nice catch!